src/main/java/com/example/cabbooking/servicetest/RideService.java

```diff
@@ -28,42 +36,54 @@ public List<RideDetailsVO> findRide(String username, LocationDetailsVO source, L
     }
 
-    public void  chooseRide(String username, String driverName) {
+    public void chooseRide(String username, String driverName) {
         DriverDetailsVO driver = driverService.getDriver(driverName);
+        Ride ride = new Ride();
         if (driver != null && driver.isAvailable()) {
 
             driver.setAvailable(false); // Mark the driver as not available
         } else {
             throw new IllegalArgumentException("Driver not found or not available");
         }
     }
-    public double calculateFare(LocationDetailsVO source,LocationDetailsVO destination){
-        double distance  = calculateDistance(source,destination);
+
+    public double calculateFare(LocationDetailsVO source, LocationDetailsVO destination) {
+        double distance = calculateDistance(source, destination);
         double distanceFare = distance * FareConstant.PER_KM_RATE;
-        double taxAmount =  distanceFare * FareConstant.TAX_RATE;
+        double taxAmount = distanceFare * FareConstant.TAX_RATE;
 
         return FareConstant.BASE_FARE + distanceFare + taxAmount + FareConstant.PLATFORM_CHARGE;
 
     }
 
     private double calculateDistance(LocationDetailsVO source, LocationDetailsVO destination) {
         double xDiff = destination.getXDistance() - source.getXDistance();
-        double yDiff = destination.getYDistance()-source.getYDistance();
+        double yDiff = destination.getYDistance() - source.getYDistance();
         return Math.sqrt(xDiff * xDiff + yDiff * yDiff);
     }
-    public void sortedDriversByDistanceToSource(LocationDetailsVO sources, List<RideDetailsVO>driverDetails){
+
+    public void sortedDriversByDistanceToSource(LocationDetailsVO sources, List<RideDetailsVO> driverDetails) {
         Collections.sort(driverDetails, new Comparator<RideDetailsVO>() {
             @Override
             public int compare(RideDetailsVO o1, RideDetailsVO o2) {
-                double distanceToO1 = calculateDistance(sources,o1.getDriver().getLocation());
-                double distanceToO2 = calculateDistance(sources,o2.getDriver().getLocation());
-                return Double.compare(distanceToO1,distanceToO2);
+                double distanceToO1 = calculateDistance(sources, o1.getDriver().getLocation());
+                double distanceToO2 = calculateDistance(sources, o2.getDriver().getLocation());
+                return Double.compare(distanceToO1, distanceToO2);
             }
         });
     }
 
+    public void cancelRide(String userName, String driverName) {
+        Ride ride = rideRepository.findByUserNameAndDriverName(userName, driverName)
+                .orElseThrow(() -> new RideNotFoundException("Ride not found for user: " + userName + " with driver: " + driverName));
+
+        if (ride.isCanceled()) {
+            throw new AlreadyCanceledException("Ride already canceled for user: " + userName + " with driver: " + driverName);
```

Comment:
please check everytime the prerequisits to for cancaling a ride

Reply:
In this case we once we cancel the ride status become true.